# Billboard Hot 100 (songs) Updated Dataframe Cleaning and Wrangling
## Based on newly released data (January 1st, 2023)

## Contents
## 01. Import libraries and datasets
## 02. Cleaning and wrangling
## 03. EDA to prepare for time series and visualization
## 04. Data export

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy
import os

In [2]:
# Create path

path = r"C:\Users\leobs\OneDrive\Desktop\LF_TS_Analysis"

In [3]:
# Import billboard hot 100 dataframe with updates

df = pd.read_csv(os.path.join(path,'01 Sourced Data', 'Main Dataframes', 'hot-100-current (3).csv'), index_col = False)

In [4]:
# Import song stream for merging song info

df_songs = pd.read_csv(os.path.join(path,'02 Manipulated Data', 'song_stream_alb_final.csv'), index_col = 0)

# 02. Cleaning

In [5]:
df.head(20)

chart_week  current_week                                        title  \
0   2022-01-01             1              All I Want For Christmas Is You   
1   2022-01-01             2            Rockin' Around The Christmas Tree   
2   2022-01-01             3                             Jingle Bell Rock   
3   2022-01-01             4                      A Holly Jolly Christmas   
4   2022-01-01             5                                   Easy On Me   
5   2022-01-01             6     It's The Most Wonderful Time Of The Year   
6   2022-01-01             7                               Last Christmas   
7   2022-01-01             8                                Feliz Navidad   
8   2022-01-01             9                                         Stay   
9   2022-01-01            10                                  Sleigh Ride   
10  2022-01-01            11  The Christmas Song (Merry Christmas To You)   
11  2022-01-01            12        Let It Snow, Let It Snow, Let It Snow   
12  2022-01-01            13                          Underneath The Tree   
13  2022-01-01            14                               Broadway Girls   
14  2022-01-01            15                                   Heat Waves   
15  2022-01-01            16            Christmas (Baby Please Come Home)   
16  2022-01-01            17  It's Beginning To Look A Lot Like Christmas   
17  2022-01-01            18                                Santa Tell Me   
18  2022-01-01            19               Rudolph The Red-Nosed Reindeer   
19  2022-01-01            20           Happy Holiday / The Holiday Season   

                                            performer  last_week  peak_pos  \
0                                        Mariah Carey        1.0         1   
1                                          Brenda Lee        2.0         2   
2                                         Bobby Helms        4.0         3   
3                                           Burl Ives        5.0         4   
4                                               Adele        3.0         1   
5                                       Andy Williams        7.0         5   
6                                               Wham!        9.0         7   
7                                      Jose Feliciano       11.0         6   
8                       The Kid LAROI & Justin Bieber        6.0         1   
9                                        The Ronettes       13.0        10   
10                                      Nat King Cole       15.0        11   
11                                        Dean Martin       17.0         8   
12                                     Kelly Clarkson       18.0        12   
13                   Lil Durk Featuring Morgan Wallen        0.0        14   
14                                      Glass Animals        8.0         7   
15                                       Darlene Love       25.0        16   
16  Perry Como And The Fontane Sisters With Mitche...       19.0        12   
17                                      Ariana Grande       24.0        17   
18                                         Gene Autry       21.0        16   
19                                      Andy Williams       22.0        18   

    wks_on_chart  
0             50  
1             44  
2             41  
3             25  
4             11  
5             26  
6             24  
7             19  
8             24  
9             15  
10            31  
11            18  
12            14  
13             1  
14            49  
15            11  
16            11  
17            12  
18            16  
19            12

In [6]:
df.tail(20)

chart_week  current_week  \
341280  2021-11-13            81   
341281  2021-11-13            82   
341282  2021-11-13            83   
341283  2021-11-13            84   
341284  2021-11-13            85   
341285  2021-11-13            86   
341286  2021-11-13            87   
341287  2021-11-13            88   
341288  2021-11-13            89   
341289  2021-11-13            90   
341290  2021-11-13            91   
341291  2021-11-13            92   
341292  2021-11-13            93   
341293  2021-11-13            94   
341294  2021-11-13            95   
341295  2021-11-13            96   
341296  2021-11-13            97   
341297  2021-11-13            98   
341298  2021-11-13            99   
341299  2021-11-13           100   

                                                title  \
341280                                     Rap Freaks   
341281                                         Eat It   
341282                                    Esta Danada   
341283                          Freedom Was A Highway   
341284                              Whole Lotta Money   
341285                               Switches & Dracs   
341286                            Half Of My Hometown   
341287                              Get Into It (Yuh)   
341288                                  Pissed Me Off   
341289                                 Lo Siento BB:/   
341290  Ya Superame (En Vivo Desde Culiacan, Sinaloa)   
341291                             Who's In Your Head   
341292                                          Money   
341293                                   Come Through   
341294                                            WFM   
341295                                Lets Go Brandon   
341296                            Just About Over You   
341297                             To Be Loved By You   
341298                               Let's Go Brandon   
341299                                     Ghost Town   

                                                performer  last_week  \
341280                                         Yung Miami        NaN   
341281                                Megan Thee Stallion        NaN   
341282                                       Ivan Cornejo       72.0   
341283                        Jimmie Allen & Brad Paisley       82.0   
341284                          BIA Featuring Nicki Minaj       75.0   
341285          Moneybagg Yo Featuring Lil Durk & EST Gee       69.0   
341286           Kelsea Ballerini Featuring Kenny Chesney       89.0   
341287                                           Doja Cat       86.0   
341288                                           Lil Durk       62.0   
341289                 Tainy, Bad Bunny & Julieta Venegas       81.0   
341290                                        Grupo Firme       92.0   
341291                                     Jonas Brothers       93.0   
341292                                               Lisa       90.0   
341293                       H.E.R. Featuring Chris Brown       99.0   
341294                                           RealestK       78.0   
341295                                     Loza Alexander       38.0   
341296                                    Priscilla Block       95.0   
341297                                    Parker McCollum       96.0   
341298  Bryson Gray Featuring Tyson James & Chandler C...       28.0   
341299                                       Benson Boone        NaN   

        peak_pos  wks_on_chart  
341280        81             1  
341281        82             1  
341282        61             5  
341283        82             4  
341284        16            17  
341285        69             2  
341286        87             2  
341287        68            13  
341288        39             3  
341289        51             4  
341290        91             3  
341291        92             4  
341292        90             2  
341293        64            17  
341294        67             4  
341295      

In [7]:
df.shape

(341300, 7)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341300 entries, 0 to 341299
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   chart_week    341300 non-null  object 
 1   current_week  341300 non-null  int64  
 2   title         341300 non-null  object 
 3   performer     341300 non-null  object 
 4   last_week     308840 non-null  float64
 5   peak_pos      341300 non-null  int64  
 6   wks_on_chart  341300 non-null  int64  
dtypes: float64(1), int64(3), object(3)
memory usage: 18.2+ MB


In [9]:
df.describe()

current_week      last_week       peak_pos   wks_on_chart
count  341300.000000  308840.000000  341300.000000  341300.000000
mean       50.500000      47.292355      40.687764       9.287767
std        28.866032      28.175538      29.333028       7.862138
min         1.000000       0.000000       1.000000       1.000000
25%        26.000000      23.000000      13.000000       4.000000
50%        51.000000      47.000000      38.000000       7.000000
75%        75.000000      71.000000      65.000000      13.000000
max       100.000000     100.000000     100.000000      91.000000

### Column renames

In [10]:
df.columns

Index(['chart_week', 'current_week', 'title', 'performer', 'last_week',
       'peak_pos', 'wks_on_chart'],
      dtype='object')

In [11]:
df.rename(columns = {'current_week': 'chart_position', 'title': 'song_title', 'performer': 'artist', 'last_week': 'prior_wk_position'}, inplace=True)

In [12]:
df.columns

Index(['chart_week', 'chart_position', 'song_title', 'artist',
       'prior_wk_position', 'peak_pos', 'wks_on_chart'],
      dtype='object')

### Duplicates

In [13]:
df_dupes = df[df.duplicated()]

In [14]:
df_dupes

Empty DataFrame
Columns: [chart_week, chart_position, song_title, artist, prior_wk_position, peak_pos, wks_on_chart]
Index: []

### Data types

In [15]:
# Check for mixed data types

for col in df.columns.tolist():
    weird = (df[[col]].applymap(type) !=df[[col]].iloc[0].apply(type)).any(axis = 1)                                                                             
    if len (df[weird]) > 0:
        print (col)

C:\Users\leobs\AppData\Local\Temp\ipykernel_59848\862644976.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) !=df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\leobs\AppData\Local\Temp\ipykernel_59848\862644976.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) !=df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\leobs\AppData\Local\Temp\ipykernel_59848\862644976.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) !=df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\leobs\AppData\Local\Temp\ipykernel_59848\862644976.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) !=df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\leobs\AppData\Local\Temp\ipykernel_59848\862644976.py:4: FutureWarn

In [16]:
df.dtypes

chart_week            object
chart_position         int64
song_title            object
artist                object
prior_wk_position    float64
peak_pos               int64
wks_on_chart           int64
dtype: object

In [17]:
# Convert the chart_week column to datetime

df['chart_week'] = pd.to_datetime(df['chart_week'])

In [18]:
df['chart_week'].dtype

dtype('<M8[ns]')

In [19]:
# Get all the unique values in the chart week columns

unique_weeks_sub = df['chart_week'].unique()

In [20]:
unique_weeks_sub.shape

(3413,)

In [21]:
df.tail(20)

chart_week  chart_position  \
341280 2021-11-13              81   
341281 2021-11-13              82   
341282 2021-11-13              83   
341283 2021-11-13              84   
341284 2021-11-13              85   
341285 2021-11-13              86   
341286 2021-11-13              87   
341287 2021-11-13              88   
341288 2021-11-13              89   
341289 2021-11-13              90   
341290 2021-11-13              91   
341291 2021-11-13              92   
341292 2021-11-13              93   
341293 2021-11-13              94   
341294 2021-11-13              95   
341295 2021-11-13              96   
341296 2021-11-13              97   
341297 2021-11-13              98   
341298 2021-11-13              99   
341299 2021-11-13             100   

                                           song_title  \
341280                                     Rap Freaks   
341281                                         Eat It   
341282                                    Esta Danada   
341283                          Freedom Was A Highway   
341284                              Whole Lotta Money   
341285                               Switches & Dracs   
341286                            Half Of My Hometown   
341287                              Get Into It (Yuh)   
341288                                  Pissed Me Off   
341289                                 Lo Siento BB:/   
341290  Ya Superame (En Vivo Desde Culiacan, Sinaloa)   
341291                             Who's In Your Head   
341292                                          Money   
341293                                   Come Through   
341294                                            WFM   
341295                                Lets Go Brandon   
341296                            Just About Over You   
341297                             To Be Loved By You   
341298                               Let's Go Brandon   
341299                                     Ghost Town   

                                                   artist  prior_wk_position  \
341280                                         Yung Miami                NaN   
341281                                Megan Thee Stallion                NaN   
341282                                       Ivan Cornejo               72.0   
341283                        Jimmie Allen & Brad Paisley               82.0   
341284                          BIA Featuring Nicki Minaj               75.0   
341285          Moneybagg Yo Featuring Lil Durk & EST Gee               69.0   
341286           Kelsea Ballerini Featuring Kenny Chesney               89.0   
341287                                           Doja Cat               86.0   
341288                                           Lil Durk               62.0   
341289                 Tainy, Bad Bunny & Julieta Venegas               81.0   
341290                                        Grupo Firme               92.0   
341291                                     Jonas Brothers               93.0   
341292                                               Lisa               90.0   
341293                       H.E.R. Featuring Chris Brown               99.0   
341294                                           RealestK               78.0   
341295                                     Loza Alexander               38.0   
341296                                    Priscilla Block               95.0   
341297                                    Parker McCollum               96.0   
341298  Bryson Gray Featuring Tyson James & Chandler C...               28.0   
341299                                       Benson Boone                NaN   

        peak_pos  wks_on_chart  
341280        81             1  
341281        82             1  
341282        61             5  
341283        82             4  
341284        16            17  
341285        69             2  
341286        87             2  
341287        68            13  
341288        39             3  
341289        51    

## Create a subset with only chart weeks from Taylor's first album onwards

In [22]:
# Create dataframe with values 10/24/2006 onward

df_sub = df.loc[(df['chart_week'] > '10/24/2006')]

In [23]:
df_sub = df_sub.sort_values(by='chart_week', ascending=False)

In [24]:
df_sub.head(10)

chart_week  chart_position                      song_title  \
10427 2023-12-30              28           I Remember Everything   
10436 2023-12-30              37                Thinkin' Bout Me   
10425 2023-12-30              26                           Water   
10426 2023-12-30              27                 Run Rudolph Run   
10428 2023-12-30              29                     Agora Hills   
10429 2023-12-30              30                      Santa Baby   
10430 2023-12-30              31               Little Saint Nick   
10431 2023-12-30              32  Please Come Home For Christmas   
10432 2023-12-30              33                      Last Night   
10433 2023-12-30              34   You're A Mean One, Mr. Grinch   

                                              artist  prior_wk_position  \
10427           Zach Bryan Featuring Kacey Musgraves               22.0   
10436                                  Morgan Wallen               32.0   
10425                                           Tyla               21.0   
10426                                    Chuck Berry               38.0   
10428                                       Doja Cat               30.0   
10429  Eartha Kitt With Henri Rene And His Orchestra               31.0   
10430                                 The Beach Boys               40.0   
10431                                         Eagles               45.0   
10432                                  Morgan Wallen               29.0   
10433                              Thurl Ravenscroft               41.0   

       peak_pos  wks_on_chart  
10427         1            17  
10436         7            42  
10425        10            12  
10426        10            21  
10428        14            13  
10429        30             8  
10430        29             7  
10431        18            14  
10432         1            47  
10433        31            11

In [25]:
df_sub.tail(20)

chart_week  chart_position                   song_title  \
260753 2006-10-28              54                       Boston   
260754 2006-10-28              55            Money In The Bank   
260755 2006-10-28              56             I Wanna Love You   
260756 2006-10-28              57         Chicken Noodle Soup.   
260757 2006-10-28              58              It Ends Tonight   
260758 2006-10-28              59                         Hurt   
260759 2006-10-28              60          Every Mile A Memory   
260761 2006-10-28              62      Everytime Tha Beat Drop   
260762 2006-10-28              63                  We Fly High   
260763 2006-10-28              64  Welcome To The Black Parade   
260764 2006-10-28              65                   Tim McGraw   
260765 2006-10-28              66                  Need A Boss   
260766 2006-10-28              67                 Tell Me Baby   
260767 2006-10-28              68              Take Me As I Am   
260768 2006-10-28              69                      Push It   
260769 2006-10-28              70                   The Riddle   
260770 2006-10-28              71      Right Where You Want Me   
260771 2006-10-28              72         Show Me What You Got   
260772 2006-10-28              73          Put Your Records On   
260773 2006-10-28              74                   Call On Me   

                                                 artist  prior_wk_position  \
260753                                        Augustana               88.0   
260754                 Lil Scrappy Featuring Young Buck               65.0   
260755                        Akon Featuring Snoop Dogg               79.0   
260756  Webstar & Young B Featuring The Voice Of Harlem               45.0   
260757                         The All-American Rejects               84.0   
260758                               Christina Aguilera               75.0   
260759                                   Dierks Bentley               58.0   
260761              Monica Featuring Dem Franchize Boyz               48.0   
260762                                        Jim Jones               87.0   
260763                              My Chemical Romance               78.0   
260764                                     Taylor Swift               68.0   
260765                      Shareefa Featuring Ludacris               67.0   
260766                            Red Hot Chili Peppers               59.0   
260767                                    Mary J. Blige               73.0   
260768                                        Rick Ross               61.0   
260769                                Five For Fighting               63.0   
260770                                  Jesse McCartney               64.0   
260771                                            JAY-Z               90.0   
260772                               Corinne Bailey Rae               71.0   
260773                                    Janet & Nelly               53.0   

        peak_pos  wks_on_chart  
260753        54             2  
260754        55             5  
260755        56             3  
260756        45             6  
260757        58             3  
260758        59             3  
260759        58            10  
260761        48            11  
260762        63             2  
260763        64             5  
260764        65             6  
260765        66            12  
260766        50             9  
260767        68             4  
260768        61             7  
260769        40            13  
260770        33             6  
260771        72             2  
260772        64            16  
260773        25            17

In [26]:
df_sub.shape

(89700, 7)

In [27]:
df_sub['artist'].value_counts(dropna=False)

artist
Taylor Swift                                     1381
Drake                                             896
Morgan Wallen                                     659
Jason Aldean                                      587
Luke Bryan                                        568
                                                 ... 
David Guetta Featuring Jennifer Hudson              1
Game Featuring Lil Wayne & Tyler, The Creator       1
Kehlani Featuring Jhene Aiko                        1
Nicole Scherzinger                                  1
Nick Lachey                                         1
Name: count, Length: 3827, dtype: int64

### Check missing values

In [28]:
df_sub.isnull().sum()

chart_week              0
chart_position          0
song_title              0
artist                  0
prior_wk_position    8834
peak_pos                0
wks_on_chart            0
dtype: int64

In [29]:
df_sub['prior_wk_position'].value_counts(dropna=False)

prior_wk_position
NaN      8834
0.0      1603
6.0       896
2.0       895
5.0       895
         ... 
96.0      505
97.0      471
98.0      449
99.0      410
100.0     380
Name: count, Length: 102, dtype: int64

In [30]:
# Fill NA with 0

df_sub['prior_wk_position'].fillna(0, inplace=True)

In [31]:
df_sub['prior_wk_position'].value_counts(dropna=False)

prior_wk_position
0.0      10437
6.0        896
7.0        895
2.0        895
5.0        895
         ...  
96.0       505
97.0       471
98.0       449
99.0       410
100.0      380
Name: count, Length: 101, dtype: int64

In [32]:
# Change data type to integer now that no NaN values

df_sub['prior_wk_position'] = df_sub['prior_wk_position'].astype(int)

In [33]:
df_sub.head(50)

chart_week  chart_position  \
10427 2023-12-30              28   
10436 2023-12-30              37   
10425 2023-12-30              26   
10426 2023-12-30              27   
10428 2023-12-30              29   
10429 2023-12-30              30   
10430 2023-12-30              31   
10431 2023-12-30              32   
10432 2023-12-30              33   
10433 2023-12-30              34   
10434 2023-12-30              35   
10435 2023-12-30              36   
10437 2023-12-30              38   
10450 2023-12-30              51   
10438 2023-12-30              39   
10440 2023-12-30              41   
10441 2023-12-30              42   
10442 2023-12-30              43   
10443 2023-12-30              44   
10444 2023-12-30              45   
10445 2023-12-30              46   
10446 2023-12-30              47   
10447 2023-12-30              48   
10448 2023-12-30              49   
10424 2023-12-30              25   
10423 2023-12-30              24   
10422 2023-12-30              23   
10421 2023-12-30              22   
10401 2023-12-30               2   
10400 2023-12-30               1   
10402 2023-12-30               3   
10403 2023-12-30               4   
10404 2023-12-30               5   
10405 2023-12-30               6   
10406 2023-12-30               7   
10407 2023-12-30               8   
10408 2023-12-30               9   
10409 2023-12-30              10   
10410 2023-12-30              11   
10411 2023-12-30              12   
10412 2023-12-30              13   
10413 2023-12-30              14   
10414 2023-12-30              15   
10415 2023-12-30              16   
10416 2023-12-30              17   
10417 2023-12-30              18   
10418 2023-12-30              19   
10419 2023-12-30              20   
10420 2023-12-30              21   
10449 2023-12-30              50   

                                              song_title  \
10427                              I Remember Everything   
10436                                   Thinkin' Bout Me   
10425                                              Water   
10426                                    Run Rudolph Run   
10428                                        Agora Hills   
10429                                         Santa Baby   
10430                                  Little Saint Nick   
10431                     Please Come Home For Christmas   
10432                                         Last Night   
10433                      You're A Mean One, Mr. Grinch   
10434                      Santa Claus Is Comin' To Town   
10435                            Wonderful Christmastime   
10437                     Rudolph The Red-Nosed Reindeer   
10450                               First Person Shooter   
10438                                          Everybody   
10440                                       Lose Control   
10441                                               FTCU   
10442                                Like It's Christmas   
10443                 Happy Holiday / The Holiday Season   
10444                                     This Christmas   
10445                                           Fast Car   
10446                                      Lil Boo Thang   
10447                           Happy Xmas (War Is Over)   
10448                               What Was I Made For?   
10424                                     Blue Christmas   
10423        It's Beginning To Look A Lot Like Christmas   
10422                                     Deck The Halls   
10421  Here Comes Santa Claus (Right Down Santa Claus...   
10401                  Rockin' Around The Christmas Tree   
10400                    All I Want For Christmas Is You   
10402                                   Jingle Bell Rock   
10403                                     Last Christmas   
10404                            A Holly Jolly Christmas   
10405                                        Lovin On Me   
10406           It's The Most Wonderful Time Of The Year   
10407    

# EDA with all artists

In [34]:
df_sub['song_title'].value_counts(dropna=False)

song_title
Stay                      172
Let It Go                 129
Rockstar                  129
Home                      125
See You Again             119
                         ... 
Livin It Up                 1
Rich N***a Shit             1
I'm Not Gonna Miss You      1
Peepin Out The Window       1
Life Is A Highway           1
Name: count, Length: 7553, dtype: int64

In [35]:
unique_artists = df_sub['artist'].nunique()

In [36]:
unique_artists

3827

In [37]:
artist_counts = df_sub['artist'].value_counts(dropna=False)

In [38]:
artist_counts.head(20)

artist
Taylor Swift        1381
Drake                896
Morgan Wallen        659
Jason Aldean         587
Luke Bryan           568
The Weeknd           537
Ed Sheeran           503
Carrie Underwood     502
Beyonce              495
Rihanna              492
Luke Combs           446
Blake Shelton        442
Imagine Dragons      442
Keith Urban          433
Kenny Chesney        418
Chris Brown          413
Katy Perry           394
P!nk                 387
Maroon 5             385
Ariana Grande        381
Name: count, dtype: int64

## Isolate Taylor only songs

In [39]:
df_sub.shape

(89700, 7)

In [40]:
# Using loc to conditionally assign a value based on whether a column includes
# string "Taylor Swift"

df_sub['result'] = 0 # Initialize the 'result' column with zeros
df_sub.loc[df_sub['artist'].str.contains('Taylor Swift'), 'result'] = 1

In [41]:
df_sub.head(15)

chart_week  chart_position                      song_title  \
10427 2023-12-30              28           I Remember Everything   
10436 2023-12-30              37                Thinkin' Bout Me   
10425 2023-12-30              26                           Water   
10426 2023-12-30              27                 Run Rudolph Run   
10428 2023-12-30              29                     Agora Hills   
10429 2023-12-30              30                      Santa Baby   
10430 2023-12-30              31               Little Saint Nick   
10431 2023-12-30              32  Please Come Home For Christmas   
10432 2023-12-30              33                      Last Night   
10433 2023-12-30              34   You're A Mean One, Mr. Grinch   
10434 2023-12-30              35   Santa Claus Is Comin' To Town   
10435 2023-12-30              36         Wonderful Christmastime   
10437 2023-12-30              38  Rudolph The Red-Nosed Reindeer   
10450 2023-12-30              51            First Person Shooter   
10438 2023-12-30              39                       Everybody   

                                              artist  prior_wk_position  \
10427           Zach Bryan Featuring Kacey Musgraves                 22   
10436                                  Morgan Wallen                 32   
10425                                           Tyla                 21   
10426                                    Chuck Berry                 38   
10428                                       Doja Cat                 30   
10429  Eartha Kitt With Henri Rene And His Orchestra                 31   
10430                                 The Beach Boys                 40   
10431                                         Eagles                 45   
10432                                  Morgan Wallen                 29   
10433                              Thurl Ravenscroft                 41   
10434                                      Jackson 5                 37   
10435                                 Paul McCartney                 35   
10437                     Gene Autry & The Pinafores                 33   
10450                        Drake Featuring J. Cole                 51   
10438             Nicki Minaj Featuring Lil Uzi Vert                 26   

       peak_pos  wks_on_chart  result  
10427         1            17       0  
10436         7            42       0  
10425        10            12       0  
10426        10            21       0  
10428        14            13       0  
10429        30             8       0  
10430        29             7       0  
10431        18            14       0  
10432         1            47       0  
10433        31            11       0  
10434        33            10       0  
10435        28            12       0  
10437        16            26       0  
10450         1            11       0  
10438        26             2       0

In [42]:
df_sub['result'].value_counts(dropna=False)

result
0    88087
1     1613
Name: count, dtype: int64

In [43]:
# Curioous about the difference being with collabs
# Will run again and use == rather than includes

df_sub['result_taylor_only'] = 0
df_sub.loc[df_sub['artist'] == 'Taylor Swift', 'result_taylor_only'] = 1

In [44]:
df_sub['result_taylor_only'].value_counts(dropna=False)

result_taylor_only
0    88319
1     1381
Name: count, dtype: int64

In [45]:
# Delineate songs that are primarily Taylors that feature other artists vs.
# songs that are from other artists and feature Taylor
# Will keep all Taylor only songs and all songs Taylor made with featured artists


df_sub['all_taylor_and_with_collabs'] = 0
df_sub.loc[(df_sub['result_taylor_only'] == 1) | (df_sub['artist'].str.contains('Taylor Swift Featuring')), 'all_taylor_and_with_collabs'] = 1

In [46]:
df_sub['all_taylor_and_with_collabs'].value_counts(dropna=False)

all_taylor_and_with_collabs
0    88176
1     1524
Name: count, dtype: int64

In [47]:
# Create a subset to show Taylor Swift collaborations

taylor_collab_songs = df_sub[(df_sub['result'] == 1) & (df_sub['result_taylor_only'] == 0)]

In [48]:
taylor_collab_songs.shape

(232, 10)

## Taylor Swift songs (only TS and with featured artists) made the Hot 100 List 1524 times since the launch of her first album, on October 24th 2006

In [49]:
# Create a billboard chart with only the Taylor Swift + collabs songs

bill_songs_ts_oct_06_dec_23 = df_sub[df_sub['all_taylor_and_with_collabs'] == 1].reset_index(drop=True)

In [50]:
bill_songs_ts_oct_06_dec_23.head(100)

chart_week  chart_position  \
0  2023-12-30              18   
1  2023-12-30              40   
2  2023-12-30              77   
3  2023-12-23              36   
4  2023-12-23              13   
..        ...             ...   
95 2023-09-02               4   
96 2023-09-02              15   
97 2023-09-02              18   
98 2023-09-02              74   
99 2023-08-26              46   

                                           song_title  \
0                                        Cruel Summer   
1   Is It Over Now? (Taylor's Version) [From The V...   
2                   You're Losing Me (From The Vault)   
3   Is It Over Now? (Taylor's Version) [From The V...   
4                                        Cruel Summer   
..                                                ...   
95                                       Cruel Summer   
96                                              Karma   
97                                          Anti-Hero   
98  I Can See You (Taylor's Version) (From The Vault)   
99                                        Blank Space   

                              artist  prior_wk_position  peak_pos  \
0                       Taylor Swift                 13         1   
1                       Taylor Swift                 36         1   
2                       Taylor Swift                 68        27   
3                       Taylor Swift                 30         1   
4                       Taylor Swift                  7         1   
..                               ...                ...       ...   
95                      Taylor Swift                  4         3   
96  Taylor Swift Featuring Ice Spice                 14         2   
97                      Taylor Swift                 20         1   
98                      Taylor Swift                 71         5   
99                      Taylor Swift                 49         1   

    wks_on_chart  result  result_taylor_only  all_taylor_and_with_collabs  
0             33       1                   1                            1  
1              8       1                   1                            1  
2              4       1                   1                            1  
3              7       1                   1                            1  
4             32       1                   1                            1  
..           ...     ...                 ...                          ...  
95            16       1                   1                            1  
96            24       1                   0                            1  
97            44       1                   1                            1  
98             7       1                   1                            1  
99            38       1                   1                            1  

[100 rows x 10 columns]

In [51]:
bill_songs_ts_oct_06_dec_23.shape

(1524, 10)

In [52]:
top_songs= bill_songs_ts_oct_06_dec_23['song_title'].value_counts(dropna=False)

In [53]:
top_songs.head(10)

song_title
Anti-Hero                   53
Shake It Off                50
You Belong With Me          50
Love Story                  49
Teardrops On My Guitar      48
Blank Space                 38
Our Song                    36
I Knew You Were Trouble.    36
Delicate                    35
Cruel Summer                33
Name: count, dtype: int64

In [54]:
bill_songs_ts_oct_06_dec_23.columns

Index(['chart_week', 'chart_position', 'song_title', 'artist',
       'prior_wk_position', 'peak_pos', 'wks_on_chart', 'result',
       'result_taylor_only', 'all_taylor_and_with_collabs'],
      dtype='object')

In [55]:
# Drop columns not needed

bill_songs_ts_oct_06_dec_23.drop(columns=['result',
       'result_taylor_only', 'all_taylor_and_with_collabs'], axis=1, inplace=True)

In [56]:
bill_songs_ts_oct_06_dec_23.columns

Index(['chart_week', 'chart_position', 'song_title', 'artist',
       'prior_wk_position', 'peak_pos', 'wks_on_chart'],
      dtype='object')

# Prepare for merge with songs df

In [57]:
df_songs.head()

song_title  \
0         Welcome To New York (Taylor's Version)   
1                 Blank Space (Taylor's Version)   
2                       Style (Taylor's Version)   
3            Out Of The Woods (Taylor's Version)   
4  All You Had To Do Was Stay (Taylor's Version)   

                       album_title song_release_date  track      acou  dance  \
0  1989 (Taylor's Version)[Deluxe]        10/27/2023      1  0.009420  0.757   
1  1989 (Taylor's Version)[Deluxe]        10/27/2023      2  0.088500  0.733   
2  1989 (Taylor's Version)[Deluxe]        10/27/2023      3  0.000421  0.511   
3  1989 (Taylor's Version)[Deluxe]        10/27/2023      4  0.000537  0.545   
4  1989 (Taylor's Version)[Deluxe]        10/27/2023      5  0.000656  0.588   

    ener    instru    live   loud  ...  alb_r_yr  album_release_date  \
0  0.610  0.000037  0.3670 -4.840  ...      2023          2023-10-27   
1  0.733  0.000000  0.1680 -5.376  ...      2023          2023-10-27   
2  0.822  0.019700  0.0899 -4.785  ...      2023          2023-10-27   
3  0.885  0.000056  0.3850 -5.968  ...      2023          2023-10-27   
4  0.721  0.000000  0.1310 -5.579  ...      2023          2023-10-27   

   w1_sales  top_alb_count  top_alb_normal alb_track_cnt tvflag  \
0   1359000              1        0.076923            22   True   
1   1359000              1        0.076923            22   True   
2   1359000              1        0.076923            22   True   
3   1359000              1        0.076923            22   True   
4   1359000              1        0.076923            22   True   

             sales_group  tot_streams_group              alb_genre  
0  High first week sales  Low total streams  Synth/Dream/Dance pop  
1  High first week sales  Low total streams  Synth/Dream/Dance pop  
2  High first week sales  Low total streams  Synth/Dream/Dance pop  
3  High first week sales  Low total streams  Synth/Dream/Dance pop  
4  High first week sales  Low total streams  Synth/Dream/Dance pop  

[5 rows x 42 columns]

In [58]:
df_songs.columns

Index(['song_title', 'album_title', 'song_release_date', 'track', 'acou',
       'dance', 'ener', 'instru', 'live', 'loud', 'speech', 'temp', 'val',
       'pop', 'dur', 'album_id', 'song_id', 'days', 'tot_streams',
       'last_stream_update', 'streams_n', 'top_song', 'top_song_flag',
       'top_s_cnt', 'tv', 'alb_num', 'album_announce_mo', 'alb_ann_day',
       'alb_ann_yr', 'album_announce_date', 'album_release_mo', 'alb_r_day',
       'alb_r_yr', 'album_release_date', 'w1_sales', 'top_alb_count',
       'top_alb_normal', 'alb_track_cnt', 'tvflag', 'sales_group',
       'tot_streams_group', 'alb_genre'],
      dtype='object')

In [59]:
df_songs.shape

(285, 42)

In [60]:
# Merge dataframes on song_title

bill_songs_merge = bill_songs_ts_oct_06_dec_23.merge(df_songs, how='left', on='song_title', indicator=True)

In [61]:
bill_songs_merge['_merge'].value_counts(dropna=False)

_merge
both          1249
left_only      275
right_only       0
Name: count, dtype: int64

In [62]:
unknown_alb= bill_songs_merge[bill_songs_merge['_merge'] == 'left_only']

In [63]:
pd.set_option('display.max_rows', 275)

In [64]:
unknown_alb

chart_week  chart_position  \
1    2023-12-30              40   
2    2023-12-30              77   
3    2023-12-23              36   
5    2023-12-23              68   
6    2023-12-16              30   
7    2023-12-16              27   
9    2023-12-16              84   
10   2023-12-09              46   
12   2023-12-09              21   
13   2023-12-09              98   
14   2023-12-09             100   
15   2023-12-09              74   
17   2023-12-02               9   
18   2023-12-02              81   
19   2023-12-02              76   
21   2023-12-02              56   
22   2023-11-25              33   
23   2023-11-25              49   
25   2023-11-25               5   
29   2023-11-25              92   
31   2023-11-25              53   
33   2023-11-18              31   
34   2023-11-18              26   
37   2023-11-18              12   
39   2023-11-18               3   
40   2023-11-18              50   
59   2023-11-11               1   
60   2023-11-11               2   
61   2023-11-11               3   
62   2023-11-11               5   
66   2023-11-11              10   
128  2023-07-22              31   
130  2023-07-22              34   
131  2023-07-22              35   
167  2023-06-17              87   
170  2023-06-10              30   
187  2023-05-06              98   
191  2023-04-29              76   
194  2023-04-22              64   
197  2023-04-15              57   
198  2023-04-08              39   
202  2023-04-01              12   
243  2022-12-10              83   
260  2022-12-03              69   
266  2022-11-26              49   
282  2022-11-19              40   
305  2022-11-12              16   
324  2022-11-05               4   
336  2022-07-09              60   
354  2022-01-08              69   
356  2022-01-01              62   
360  2021-12-25              81   
361  2021-12-18              74   
365  2021-12-11              82   
371  2021-12-04              70   
372  2021-11-27              63   
375  2021-11-27              66   
390  2021-11-27              22   
394  2021-11-27              47   
397  2021-11-27              43   
399  2021-05-08              61   
401  2021-05-01              60   
407  2021-04-24              62   
410  2021-04-24              72   
412  2021-04-17              60   
413  2021-04-10              51   
414  2021-04-10              59   
415  2021-04-03              50   
416  2021-03-27              39   
417  2021-03-20              42   
418  2021-03-13              32   
419  2021-03-06              29   
420  2021-02-27              25   
422  2021-02-20              29   
423  2021-02-13              29   
424  2021-02-06              28   
425  2021-01-30              28   
426  2021-01-23              29   
427  2021-01-16              22   
428  2021-01-09              23   
429  2021-01-02              38   
430  2021-01-02              91   
431  2021-01-02              97   
432  2021-01-02              77   
433  2020-12-26              34   
434  2020-12-26              39   
435  2020-12-26              40   
436  2020-12-26              45   
437  2020-12-26               1   
438  2020-12-26              21   
439  2020-12-26              82   
440  2020-12-26              52   
441  2020-12-26              75   
442  2020-12-26              63   
443  2020-12-26              57   
444  2020-12-26              61   
445  2020-12-26              71   
446  2020-12-26              68   
447  2020-12-26              67   
448  2020-11-07              96   
449  2020-10-31              88   
450  2020-10-24              85   
451  2020-10-17              77   
452  2020-10-10              53   
453  2020-10-03              49   
454  2020-09-26              52   
455  2020-09-19              52   
456  2020-09-12              45   
457  2020-09-12              92   
458  2020-09-05              36   
459  2020-09-05              85   
460  2020-09-05              98   
461  2020-08-29              33   
462  2020-08-29

In [65]:
unique_songs_unknown = unknown_alb['song_title'].value_counts()

In [66]:
unique_songs_unknown

song_title
Teardrops On My Guitar                                          48
ME!                                                             20
Willow                                                          20
Today Was A Fairytale                                           18
Safe & Sound                                                    17
Cardigan                                                        14
Eyes Open                                                       13
Snow On The Beach                                                8
Crazier                                                          8
Is It Over Now? (Taylor's Version) [From The Vault]              8
The 1                                                            6
Now That We Don't Talk (Taylor's Version) [From The Vault]       6
All Of The Girls You Loved Before                                6
Christmas Tree Farm (Old Timey Version)                          5
Slut! (Taylor's Version) [From The Vault]          

In [67]:
unknown_alb.shape

(275, 49)

In [68]:
bill_songs_merge['song_title'].value_counts()

song_title
Anti-Hero                                                       53
Shake It Off                                                    50
You Belong With Me                                              50
Love Story                                                      49
Teardrops On My Guitar                                          48
Blank Space                                                     38
Our Song                                                        36
I Knew You Were Trouble.                                        36
Delicate                                                        35
Cruel Summer                                                    33
Style                                                           32
Karma                                                           30
Lavender Haze                                                   29
Wildest Dreams                                                  27
Bad Blood                                          

### Edit formatting for bill_songs_ts_oct_06_dec_23 to retry merge

In [69]:
# Display unique values in song_title

unique_songs = bill_songs_ts_oct_06_dec_23['song_title'].unique()

In [70]:
unique_songs

array(['Cruel Summer',
       "Is It Over Now? (Taylor's Version) [From The Vault]",
       "You're Losing Me (From The Vault)",
       "Now That We Don't Talk (Taylor's Version) [From The Vault]",
       "You're Losing Me", "Slut! (Taylor's Version) [From The Vault]",
       "Say Don't Go (Taylor's Version) [From The Vault]",
       "Style (Taylor's Version)", "Bad Blood (Taylor's Version)",
       "Out Of The Woods (Taylor's Version)",
       "Wildest Dreams (Taylor's Version)",
       "Suburban Legends (Taylor's Version) [From The Vault]",
       "Blank Space (Taylor's Version)",
       "Shake It Off (Taylor's Version)",
       "All You Had To Do Was Stay (Taylor's Version)",
       "New Romantics (Taylor's Version)", "Clean (Taylor's Version)",
       "Welcome To New York (Taylor's Version)",
       "I Know Places (Taylor's Version)",
       "I Wish You Would (Taylor's Version)",
       "Wonderland (Taylor's Version)",
       "How You Get The Girl (Taylor's Version)",
       "This 

In [71]:
unique_songs.shape

(226,)

In [72]:
# Remove non-studio album songs (8)

songs_to_drop = ['Breathless', 'Carolina', 'Christmas Tree Farm (Old Timey Version)', 'Only The Young', 'Christmas Tree Farm', 
    'Sweeter Than Fiction', 'Eyes Open', 'Crazier']

In [73]:
bill_songs_ts_oct_06_dec_23 = bill_songs_ts_oct_06_dec_23[~bill_songs_ts_oct_06_dec_23['song_title'].isin(songs_to_drop)]

In [74]:
unique_songs2 = bill_songs_ts_oct_06_dec_23['song_title'].unique()

In [75]:
unique_songs2.shape

(218,)

In [76]:
# Replace capital values with lowercase for evermore and folklore albums to match for merge

ever_folk_values_to_change = {'Willow': 'willow', 'Champagne Problems': 'champagne problems', "'Tis The Damn Season": "'tis the damn season'", 'Gold Rush': 'gold rush',
       'Tolerate It': 'tolerate it', 'Happiness': 'happiness', 'Ivy': 'ivy', 'Dorothea': 'dorthea', 'Long Story Short': 'long story short',
       'Cowboy Like Me': 'cowboy like me', 'Marjorie': 'marjorie', 'Closure': 'closure', 'Cardigan': 'cardigan', 'The 1': 'the 1',
       'The Last Great American Dynasty':'the last great american dynasty', 'My Tears Ricochet': 'my tears ricochet', 'August': 'august','Invisible String': 'invisible string', 'Betty': 'betty', "Mirrorball":"mirrorball", 'Seven': 'seven',
       'This Is Me Trying': 'this is me trying', 'Illicit Affairs': 'illicit affairs', 'Mad Woman': 'mad woman', 'Epiphany': 'epiphany', 'Peace': 'peace', 'Hoax': 'hoax'}


In [77]:
ever_folk_values_to_change

# 27 changes

{'Willow': 'willow',
 'Champagne Problems': 'champagne problems',
 "'Tis The Damn Season": "'tis the damn season'",
 'Gold Rush': 'gold rush',
 'Tolerate It': 'tolerate it',
 'Happiness': 'happiness',
 'Ivy': 'ivy',
 'Dorothea': 'dorthea',
 'Long Story Short': 'long story short',
 'Cowboy Like Me': 'cowboy like me',
 'Marjorie': 'marjorie',
 'Closure': 'closure',
 'Cardigan': 'cardigan',
 'The 1': 'the 1',
 'The Last Great American Dynasty': 'the last great american dynasty',
 'My Tears Ricochet': 'my tears ricochet',
 'August': 'august',
 'Invisible String': 'invisible string',
 'Betty': 'betty',
 'Mirrorball': 'mirrorball',
 'Seven': 'seven',
 'This Is Me Trying': 'this is me trying',
 'Illicit Affairs': 'illicit affairs',
 'Mad Woman': 'mad woman',
 'Epiphany': 'epiphany',
 'Peace': 'peace',
 'Hoax': 'hoax'}

In [78]:
bill_songs_ts_oct_06_dec_23['song_title'].replace(ever_folk_values_to_change, inplace=True)

In [79]:
unique_songs3 = bill_songs_ts_oct_06_dec_23['song_title'].unique()

In [80]:
# Recheck unique songs with lowercase shift

unique_songs3

array(['Cruel Summer',
       "Is It Over Now? (Taylor's Version) [From The Vault]",
       "You're Losing Me (From The Vault)",
       "Now That We Don't Talk (Taylor's Version) [From The Vault]",
       "You're Losing Me", "Slut! (Taylor's Version) [From The Vault]",
       "Say Don't Go (Taylor's Version) [From The Vault]",
       "Style (Taylor's Version)", "Bad Blood (Taylor's Version)",
       "Out Of The Woods (Taylor's Version)",
       "Wildest Dreams (Taylor's Version)",
       "Suburban Legends (Taylor's Version) [From The Vault]",
       "Blank Space (Taylor's Version)",
       "Shake It Off (Taylor's Version)",
       "All You Had To Do Was Stay (Taylor's Version)",
       "New Romantics (Taylor's Version)", "Clean (Taylor's Version)",
       "Welcome To New York (Taylor's Version)",
       "I Know Places (Taylor's Version)",
       "I Wish You Would (Taylor's Version)",
       "Wonderland (Taylor's Version)",
       "How You Get The Girl (Taylor's Version)",
       "This 

In [81]:
# Need to reformat the 1989TV tracks- replace [] with () for 1989 TV songs

bill_songs_ts_oct_06_dec_23['song_title'].replace({"Now That We Don't Talk (Taylor's Version) [From The Vault]": "Now That We Don't Talk (Taylor's Version)(From The Vault)", 
                           "Slut! (Taylor's Version) [From The Vault]": '''"Slut!"(Taylor's Version)(From The Vault)''', 
                           "Say Don't Go (Taylor's Version) [From The Vault]": "Say Don't Go (Taylor's Version)(From The Vault)", 
                           "Suburban Legends (Taylor's Version) [From The Vault]": "Suburban Legends (Taylor's Version)(From The Vault)"}, 
                                                  inplace=True)

In [82]:
unique_songs_unknown

song_title
Teardrops On My Guitar                                          48
ME!                                                             20
Willow                                                          20
Today Was A Fairytale                                           18
Safe & Sound                                                    17
Cardigan                                                        14
Eyes Open                                                       13
Snow On The Beach                                                8
Crazier                                                          8
Is It Over Now? (Taylor's Version) [From The Vault]              8
The 1                                                            6
Now That We Don't Talk (Taylor's Version) [From The Vault]       6
All Of The Girls You Loved Before                                6
Christmas Tree Farm (Old Timey Version)                          5
Slut! (Taylor's Version) [From The Vault]          

In [83]:
# Remove 'Ronan', 'You're Losing Me (From The Vault)', 'All Of The Girls You Loved Before'
# since they are single releases
# 3 songs

singles_remove = ['Ronan', "You're Losing Me (From The Vault)", 'All Of The Girls You Loved Before']

In [84]:
bill_songs_ts_oct_06_dec_23 = bill_songs_ts_oct_06_dec_23[~bill_songs_ts_oct_06_dec_23['song_title'].isin(singles_remove)]

In [85]:
# Check 
unique_songs4 = bill_songs_ts_oct_06_dec_23['song_title'].unique()

In [86]:
unique_songs4

array(['Cruel Summer',
       "Is It Over Now? (Taylor's Version) [From The Vault]",
       "Now That We Don't Talk (Taylor's Version)(From The Vault)",
       "You're Losing Me", '"Slut!"(Taylor\'s Version)(From The Vault)',
       "Say Don't Go (Taylor's Version)(From The Vault)",
       "Style (Taylor's Version)", "Bad Blood (Taylor's Version)",
       "Out Of The Woods (Taylor's Version)",
       "Wildest Dreams (Taylor's Version)",
       "Suburban Legends (Taylor's Version)(From The Vault)",
       "Blank Space (Taylor's Version)",
       "Shake It Off (Taylor's Version)",
       "All You Had To Do Was Stay (Taylor's Version)",
       "New Romantics (Taylor's Version)", "Clean (Taylor's Version)",
       "Welcome To New York (Taylor's Version)",
       "I Know Places (Taylor's Version)",
       "I Wish You Would (Taylor's Version)",
       "Wonderland (Taylor's Version)",
       "How You Get The Girl (Taylor's Version)",
       "This Love (Taylor's Version)",
       "You Are In L

In [87]:
# Rename 'Teardrops On My Guitar'

bill_songs_ts_oct_06_dec_23['song_title'].replace({'Teardrops On My Guitar':"Teardrops On My Guitar - Radio Single Remix"}, inplace=True)

In [88]:
# Need to reformat songs that have featured artists:

bill_songs_ts_oct_06_dec_23['song_title'].replace({
    "Castles Crumbling (Taylor's Version) (From The Vault)":"Castles Crumbling (feat. Hayley Williams) (Taylor's Version) (From The Vault)" ,
    "Electric Touch (Taylor's Version) (From The Vault)":"Electric Touch (feat. Fall Out Boy) (Taylor's Version) (From The Vault)",
    "Everything Has Changed (Taylor's Version)":"Everything Has Changed (feat. Ed Sheeran) (Taylor's Version)",
    "I Bet You Think About Me (Taylor's Version) (From The Vault)":"I Bet You Think About Me (feat. Chris Stapleton) (Taylor's Version) (From The Vault)",
    "Nothing New (Taylor's Version) (From The Vault)":"Nothing New (feat. Phoebe Bridgers) (Taylor's Version) (From The Vault)",
    "Snow On The Beach":"Snow On The Beach (feat. Lana Del Rey)",
    "Soon You'll Get Better":"Soon You'll Get Better (feat. The Chicks)",
    "The Last Time (Taylor's Version)":"The Last Time (feat. Gary Lightbody of Snow Patrol) (Taylor's Version)"}
     ,inplace=True)

### Changes tracker
8 removals due to non studio
27 lowercase because of folklore/evermore
4 1989 TV title shifts
3 removals due to singles
1 rename Teardrops
8 rename due to featured artists

---
51 changes


## Re-attempt merge with updates

In [89]:
bill_songs_ts_oct_06_dec_23.shape

(1482, 7)

In [90]:
# Count of songs
unique_songs5 = bill_songs_ts_oct_06_dec_23['song_title'].unique()

In [91]:
unique_songs5.shape

(215,)

In [92]:
df_songs.shape

(285, 42)

In [93]:
# Merge on song title, with updates

bill_songs_merge2 = bill_songs_ts_oct_06_dec_23.merge(df_songs, how='left', on='song_title', indicator=True)

In [94]:
bill_songs_merge2.shape

(1482, 49)

In [95]:
bill_songs_merge2.columns

Index(['chart_week', 'chart_position', 'song_title', 'artist',
       'prior_wk_position', 'peak_pos', 'wks_on_chart', 'album_title',
       'song_release_date', 'track', 'acou', 'dance', 'ener', 'instru', 'live',
       'loud', 'speech', 'temp', 'val', 'pop', 'dur', 'album_id', 'song_id',
       'days', 'tot_streams', 'last_stream_update', 'streams_n', 'top_song',
       'top_song_flag', 'top_s_cnt', 'tv', 'alb_num', 'album_announce_mo',
       'alb_ann_day', 'alb_ann_yr', 'album_announce_date', 'album_release_mo',
       'alb_r_day', 'alb_r_yr', 'album_release_date', 'w1_sales',
       'top_alb_count', 'top_alb_normal', 'alb_track_cnt', 'tvflag',
       'sales_group', 'tot_streams_group', 'alb_genre', '_merge'],
      dtype='object')

In [96]:
bill_songs_merge2['_merge'].value_counts(dropna=False)

_merge
both          1383
left_only       99
right_only       0
Name: count, dtype: int64

In [97]:
unknown_alb_2 = bill_songs_merge2[bill_songs_merge2['_merge'] == 'left_only']

In [98]:
unknown_alb_2 = unknown_alb_2['song_title']

In [99]:
unknown_alb_2

1       Is It Over Now? (Taylor's Version) [From The V...
2       Is It Over Now? (Taylor's Version) [From The V...
4       Is It Over Now? (Taylor's Version) [From The V...
6       Now That We Don't Talk (Taylor's Version)(From...
7                                        You're Losing Me
9       Is It Over Now? (Taylor's Version) [From The V...
10              "Slut!"(Taylor's Version)(From The Vault)
11        Say Don't Go (Taylor's Version)(From The Vault)
12      Now That We Don't Talk (Taylor's Version)(From...
14      Is It Over Now? (Taylor's Version) [From The V...
15        Say Don't Go (Taylor's Version)(From The Vault)
16              "Slut!"(Taylor's Version)(From The Vault)
18      Now That We Don't Talk (Taylor's Version)(From...
19      Now That We Don't Talk (Taylor's Version)(From...
20              "Slut!"(Taylor's Version)(From The Vault)
22      Is It Over Now? (Taylor's Version) [From The V...
26      Suburban Legends (Taylor's Version)(From The V...
28        Say 

Drop: "You're Losing Me", "Safe & Sound", "Today Was A Fairytale"

Rename: "'tis the damn season'": "'tis the damn season", "Exhile":"exile",
"Evermore":"evermore", "Coney Island":"coney island", "ME !": "ME! (feat. Brendon Urie of Panic! At The Disco)", "Run (Taylor's Version) (From The Vault)": "Run (feat. Ed Sheeran) (Taylor's Version) (From The Vault)"


Issues:
- x You're Losing Me
- X Me!
- xToday Was A Fairytale
- X Exile
- X Evermore
- X Coney Island
- X 'tis the damn season'
- x "Run (Taylor's Version) (From The Vault)"

Start here, will revisit Taylor's Version issues after

In [100]:
drop_again = ["You're Losing Me", "Safe & Sound", "Today Was A Fairytale"]

In [101]:
bill_songs_ts_oct_06_dec_23 = bill_songs_ts_oct_06_dec_23[~bill_songs_ts_oct_06_dec_23['song_title'].isin(drop_again)]

In [102]:
# Renames

bill_songs_ts_oct_06_dec_23['song_title'].replace({
                            "'tis the damn season'": "'tis the damn season", 
                            "Exhile":"exile",
                            "Evermore":"evermore", 
                            "Coney Island":"coney island", 
                            "ME!": "ME! (feat. Brendon Urie of Panic! At The Disco)",
                            "Run (Taylor's Version) (From The Vault)": "Run (feat. Ed Sheeran) (Taylor's Version) (From The Vault)"
                             }, inplace=True)

- Renames based on formatting () and spaces

'"Slut!"(Taylor\'s Version)(From The Vault)':'"Slut!" (Taylor\'s Version) (From The Vault)',
"Is It Over Now? (Taylor's Version) [From The Vault]": "Is It Over Now (Taylor's Version) (From The Vault)",
"Now That We Don't Talk (Taylor's Version)(From the Vault)":"Now That We Don't Talk (Taylor's Version) (From the Vault)",
"Say Don't Go (Taylor's Version)(From The Vault)":"Say Don't Go (Taylor's Version) (From The Vault)",
"Suburban Legends (Taylor's Version)(From The Vault)":"Suburban Legends (Taylor's Version) (From The Vault)",
"No Body, No Crime": "no body, no crime (feat. HAIM)"

In [103]:
# Renames based on formatting () and spaces

bill_songs_ts_oct_06_dec_23['song_title'].replace({
    '"Slut!"(Taylor\'s Version)(From The Vault)': '"Slut!" (Taylor\'s Version) (From The Vault)',
"Is It Over Now? (Taylor's Version) [From The Vault]": "Is It Over Now (Taylor's Version) (From The Vault)",
"Now That We Don't Talk (Taylor's Version)(From the Vault)":"Now That We Don't Talk (Taylor's Version) (From the Vault)",
"Say Don't Go (Taylor's Version)(From The Vault)":"Say Don't Go (Taylor's Version) (From The Vault)",
"Suburban Legends (Taylor's Version)(From The Vault)":"Suburban Legends (Taylor's Version) (From The Vault)",
"No Body, No Crime": "no body, no crime (feat. HAIM)"},
    inplace=True)

### Attempt merge again

In [104]:
# Merge on song title, with updates

bill_songs_merge3 = bill_songs_ts_oct_06_dec_23.merge(df_songs, how='left', on='song_title', indicator=True)

In [105]:
bill_songs_merge3['_merge'].value_counts(dropna=False)

_merge
both          1419
left_only       27
right_only       0
Name: count, dtype: int64

In [106]:
unknown_alb_3 = bill_songs_merge3[bill_songs_merge3['_merge'] == 'left_only']

In [107]:
unknown_alb_3['song_title'].value_counts(dropna=False)

song_title
Is It Over Now (Taylor's Version) (From The Vault)             8
Now That We Don't Talk (Taylor's Version)(From The Vault)      6
Exile                                                          5
You All Over Me (Taylor's Version) (From The Vault)            2
'tis the damn season                                           2
When Emma Falls In Love (Taylor's Version) (From The Vault)    1
coney island                                                   1
evermore                                                       1
dorthea                                                        1
Name: count, dtype: int64

In [108]:
unknown_alb_3['song_title'].unique()

array(["Is It Over Now (Taylor's Version) (From The Vault)",
       "Now That We Don't Talk (Taylor's Version)(From The Vault)",
       "When Emma Falls In Love (Taylor's Version) (From The Vault)",
       "You All Over Me (Taylor's Version) (From The Vault)",
       "'tis the damn season", 'coney island', 'evermore', 'dorthea',
       'Exile'], dtype=object)

In [109]:
# Make more formatting shifts 

bill_songs_ts_oct_06_dec_23['song_title'].replace({
                            "Is It Over Now (Taylor's Version) (From The Vault)": "Is It Over Now? (Taylor's Version) (From The Vault)",
                            "Now That We Don't Talk (Taylor's Version)(From The Vault)":"Now That We Don't Talk (Taylor's Version) (From The Vault)",
                            "When Emma Falls In Love (Taylor's Version) (From The Vault)": "When Emma Falls in Love (Taylor's Version) (From The Vault)",
                            "You All Over Me (Taylor's Version) (From The Vault)": "You All Over Me (feat. Maren Morris) (Taylor's Version) (From The Vault)",
                            "'tis the damn season": "tis the damn season",
                            "coney island": "coney island (feat. The National)",
                            'evermore': "evermore (feat. Bon Iver)",
                            "dorthea": "dorothea",
                            "Exile": "exile (feat. Bon Iver)"},
                        inplace=True)

## Merge attempt 4

In [110]:
# Merge on song title, with updates

bill_songs_merge4 = bill_songs_ts_oct_06_dec_23.merge(df_songs, how='left', on='song_title', indicator=True)

In [111]:
bill_songs_merge4['_merge'].value_counts(dropna=False)

_merge
both          1444
left_only        2
right_only       0
Name: count, dtype: int64

In [112]:
last_issues = bill_songs_merge4[bill_songs_merge4['_merge'] == 'left_only']

In [113]:
last_issues

chart_week  chart_position           song_title        artist  \
414 2021-01-02              91  tis the damn season  Taylor Swift   
418 2020-12-26              39  tis the damn season  Taylor Swift   

     prior_wk_position  peak_pos  wks_on_chart album_title song_release_date  \
414                 39        39             2         NaN               NaN   
418                  0        39             1         NaN               NaN   

     track  ...  album_release_date  w1_sales  top_alb_count  top_alb_normal  \
414    NaN  ...                 NaN       NaN            NaN             NaN   
418    NaN  ...                 NaN       NaN            NaN             NaN   

     alb_track_cnt  tvflag  sales_group  tot_streams_group  alb_genre  \
414            NaN     NaN          NaN                NaN        NaN   
418            NaN     NaN          NaN                NaN        NaN   

        _merge  
414  left_only  
418  left_only  

[2 rows x 49 columns]

In [114]:
df_songs.columns

Index(['song_title', 'album_title', 'song_release_date', 'track', 'acou',
       'dance', 'ener', 'instru', 'live', 'loud', 'speech', 'temp', 'val',
       'pop', 'dur', 'album_id', 'song_id', 'days', 'tot_streams',
       'last_stream_update', 'streams_n', 'top_song', 'top_song_flag',
       'top_s_cnt', 'tv', 'alb_num', 'album_announce_mo', 'alb_ann_day',
       'alb_ann_yr', 'album_announce_date', 'album_release_mo', 'alb_r_day',
       'alb_r_yr', 'album_release_date', 'w1_sales', 'top_alb_count',
       'top_alb_normal', 'alb_track_cnt', 'tvflag', 'sales_group',
       'tot_streams_group', 'alb_genre'],
      dtype='object')

In [115]:
tis_check = df_songs[df_songs['album_id'] == "EVER"]

In [116]:
tis_check

song_title                album_title  \
123                                 willow  evermore (deluxe version)   
124                     champagne problems  evermore (deluxe version)   
125                              gold rush  evermore (deluxe version)   
126                   ‘tis the damn season  evermore (deluxe version)   
127                            tolerate it  evermore (deluxe version)   
128         no body, no crime (feat. HAIM)  evermore (deluxe version)   
129                              happiness  evermore (deluxe version)   
130                               dorothea  evermore (deluxe version)   
131      coney island (feat. The National)  evermore (deluxe version)   
132                                    ivy  evermore (deluxe version)   
133                         cowboy like me  evermore (deluxe version)   
134                       long story short  evermore (deluxe version)   
135                               marjorie  evermore (deluxe version)   
136                                closure  evermore (deluxe version)   
137              evermore (feat. Bon Iver)  evermore (deluxe version)   
138  right where you left me - bonus track  evermore (deluxe version)   
139          it's time to go - bonus track  evermore (deluxe version)   

    song_release_date  track   acou  dance   ener    instru    live    loud  \
123        12/11/2020      1  0.833  0.392  0.574  0.001790  0.1450  -9.195   
124        12/11/2020      2  0.920  0.462  0.240  0.000000  0.1130 -12.077   
125        12/11/2020      3  0.830  0.512  0.462  0.166000  0.1210 -10.491   
126        12/11/2020      4  0.735  0.575  0.434  0.000066  0.1050  -8.193   
127        12/11/2020      5  0.878  0.316  0.361  0.000027  0.0797 -10.381   
128        12/11/2020      6  0.418  0.546  0.613  0.000000  0.1030  -7.589   
129        12/11/2020      7  0.870  0.559  0.334  0.000000  0.1140 -10.733   
130        12/11/2020      8  0.696  0.605  0.488  0.000000  0.1290  -8.322   
131        12/11/2020      9  0.819  0.537  0.537  0.000904  0.1420 -11.266   
132        12/11/2020     10  0.855  0.515  0.545  0.000020  0.0921  -9.277   
133        12/11/2020     11  0.768  0.604  0.517  0.000155  0.1230  -9.014   
134        12/11/2020     12  0.660  0.546  0.730  0.179000  0.0972  -7.704   
135        12/11/2020     13  0.876  0.535  0.561  0.000136  0.1150 -11.609   
136        12/11/2020     14  0.835  0.689  0.704  0.000005  0.1340 -10.813   
137        12/11/2020     15  0.937  0.390  0.270  0.002270  0.1110 -10.673   
138          1/7/2021     16  0.769  0.581  0.619  0.000106  0.1170  -6.524   
139          1/7/2021     17  0.801  0.592  0.410  0.000014  0.0900 -12.426   

     ...  alb_r_yr  album_release_date  w1_sales  top_alb_count  \
123  ...      2020          2020-12-11    154500            151   
124  ...      2020          2020-12-11    154500            151   
125  ...      2020          2020-12-11    154500            151   
126  ...      2020          2020-12-11    154500            151   
127  ...      2020          2020-12-11    154500            151   
128  ...      2020          2020-12-11    154500            151   
129  ...      2020          2020-12-11    154500            151   
130  ...      2020          2020-12-11    154500            151   
131  ...      2020          2020-12-11    154500            151   
132  ...      2020          2020-12-11    154500            151   
133  ...      2020          2020-12-11    154500            151   
134  ...      2020          2020-12-11    154500            151   
135  ...      2020          2020-12-11    154500            151   
136  ...      2020          2020-12-11    154500            151   
137  ...      2020          2020-12-11    154500            151   
138  ...      2020          2020-12-11    154500            151   
139  ...      2020          2020-12-11    154500            151   

     top_alb_normal alb_track_cnt tvflag           sales_group  \
123        0.142051  

In [117]:
# Make more formatting shifts 

bill_songs_ts_oct_06_dec_23['song_title'].replace({
                            "tis the damn season":"‘tis the damn season"},
                        inplace=True)

In [118]:
# Merge on song title, with updates

bill_songs_merge5 = bill_songs_ts_oct_06_dec_23.merge(df_songs, how='left', on='song_title', indicator=True)

In [119]:
bill_songs_merge5['_merge'].value_counts(dropna=False)

_merge
both          1446
left_only        0
right_only       0
Name: count, dtype: int64

## Succesfully merged billboard song data with all song information!
#### Exported below

# Data export

In [120]:
# Export billboard 100 with all artists from Taylor's career time window
# Note: some formatting issues for Taylor's songs (corrected in below df)

df_sub.to_csv(os.path.join(path, '02 Manipulated Data', 'df_hot_100_all_artists_dec_23.csv'))

In [121]:
# Export billboard 100 songs with Taylor Swift only, during her career
# Includes all song information

bill_songs_merge5.to_csv(os.path.join(path, '02 Manipulated Data', 'df_hot_100_ts_only_dec_23.csv'))